api/src/main/java/org/apache/iceberg/SortOrder.java

```diff
@@ -0,0 +1,212 @@
+  private static final SortOrder UNSORTED_ORDER = new SortOrder(new Schema(), 0, Collections.emptyList());
+
+  private final Schema schema;
+  private final int orderId;
+  private final SortField[] fields;
+
+  private transient volatile List<SortField> fieldList;
+
+  private SortOrder(Schema schema, int orderId, List<SortField> fields) {
+    this.schema = schema;
+    this.orderId = orderId;
+    this.fields = fields.toArray(new SortField[0]);
+  }
+
+  public Schema schema() {
+    return schema;
+  }
+
+  public int orderId() {
+    return orderId;
+  }
+
+  public List<SortField> fields() {
+    return lazyFieldList();
+  }
+
+  public boolean isUnsorted() {
+    return fields.length < 1;
+  }
+
+  public boolean satisfies(SortOrder anotherSortOrder) {
+    // any ordering satisfies an unsorted ordering
+    if (anotherSortOrder.isUnsorted()) {
+      return true;
+    }
+
+    // this ordering cannot satisfy an ordering with more sort fields
+    if (anotherSortOrder.fields.length > fields.length) {
+      return false;
+    }
+
+    // this ordering has either more or the same number of sort fields
+    return IntStream.range(0, anotherSortOrder.fields.length)
+        .allMatch(index -> fields[index].equals(anotherSortOrder.fields[index]));
+  }
+
+  public boolean sameOrder(SortOrder anotherSortOrder) {
+    return Arrays.equals(fields, anotherSortOrder.fields);
+  }
+
+  private List<SortField> lazyFieldList() {
+    if (fieldList == null) {
+      synchronized (this) {
+        if (fieldList == null) {
+          this.fieldList = ImmutableList.copyOf(fields);
+        }
+      }
+    }
+    return fieldList;
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("[");
+    for (SortField field : fields) {
+      sb.append("\n");
+      sb.append("  ").append(field);
+    }
+    if (fields.length > 0) {
+      sb.append("\n");
+    }
+    sb.append("]");
+    return sb.toString();
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    } else if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+
+    SortOrder that = (SortOrder) other;
+    return orderId == that.orderId && Arrays.equals(fields, that.fields);
+  }
+
+  @Override
+  public int hashCode() {
+    return 31 * Integer.hashCode(orderId) + Arrays.hashCode(fields);
+  }
+
+  public static SortOrder unsorted() {
+    return UNSORTED_ORDER;
+  }
+
+  public static Builder builderFor(Schema schema) {
+    return new Builder(schema);
+  }
+
+  public static class Builder {
+    private final Schema schema;
+    private final List<SortField> fields = Lists.newArrayList();
+    private int orderId = 0;
+    private boolean caseSensitive = true;
+
+    private Builder(Schema schema) {
+      this.schema = schema;
+    }
+
+    public Builder asc(String name, NullOrder nullOrder) {
```

Comment:
I did not default the null order as the SQL standard does not define this. Spark and Presto, as well as RDBMS, use different default null orderings.

Reply:
I think even if other frameworks don't define this it's ok to have a default? Like If a user doesn't care then we should just make the decision for them right?

Reply:
Yeah, I'll be fine with picking a default value ourselves.

Reply:
What does Presto use? That seems like a reasonable choice. I'd like to avoid requiring field-level `NullOrder` when using this API.

That might be slightly annoying, since the builder pattern shouldn't rely on calling methods in a particular order. That is, I think that these two should produce the same sort order:

```java
// ORDER BY x ASC NULLS FIRST, y DESC NULLS FIRST
SortOrder.builderFor(schema)
    .asc("x")
    .desc("y")
    .nullsFirst()
    .build();

// ORDER BY x ASC NULLS FIRST, y DESC NULLS FIRST
SortOrder.builderFor(schema)
    .nullsFirst()
    .asc("x")
    .desc("y")
    .build();
```

Reply:
Presto seems to behave differently than any other system. I wanted to follow Spark, Hive and MySQL until I saw a stale PR in Hive to switch to NULLS LAST by default. Having said that, I still propose to match Spark and MySQL and also allow setting a null order per column as that is possible in many systems.

Reply:
That sounds good to me for the defaults. I'd still like to make it easy to use the default if you don't care, and easy to set the default for this sort order, like the examples above.

Reply:
How commonly do we expect to see cases when users want to configure a null order for all columns through a global param? In most cases, the ordering will come from query engines in a form of expressions that have a null order per column. If users create a sort order manually and don't care, they can just skip null orders. I do get the use case of being able to set the default but my worry that it complicates the implementation. We will need to defer assigning a null order in asc/desc and iterate through sort columns in `build` and see whether we already assigned the null order for a particular column.

Reply:
Yeah, that's a fair argument. We can always add `nullsFirst()` and `nullsLast()` calls later and keep just the `asc` and `desc` calls for now.

Reply:
Presto does `NULLS LAST` by default in both directions because we expect that users don't want to see nulls. Consider these queries:
```sql
SELECT x FROM t ORDER BY x LIMIT 10;
SELECT x FROM t ORDER BY x DESC limit 10;
```
In both of these cases, the user likely wants to see some data, not a bunch of nulls. This behavior is allowed by the SQL standard:

> If `<null ordering>` is not specified, then an implementation-defined `<null ordering>` is implicit. The implementation-defined default for `<null ordering>` shall not depend on the context outside of `<sort specification list>`.